_posts/2019-03-21-timezone-in-rails-and-javascript.md

```diff
@@ -0,0 +1,96 @@
+---
+layout: post
+title:  "Timezone in Rails and JavaScript"
+date:   2019-03-21 4:00:00 -0400
+categories: Ruby on Rails, JavaScript, timezone, Date
+author: Jason Ji
+---
+
+This blog contains two parts:
+  1. Timezone in Rails App
+  2. Timezone in the browser
+
+-----
+<br>
+
+## Timezone in Rails
+<div class="information information-tip">
+  Rails provides great tools for working with time zones but there's still a lot of things that can go wrong. This post aims to shed some light on these gotchas and provide solutions to the most common problems.
+</div>
+
+Rails fools you to believe it got you covered all the time, but I've learnt the hard way that I can't get away without knowing when and how Rails is helping me. There are so many timezone factors we need to consider:
+  - database
+  - server
+  - dev machine
+  - system configured
+  - user specific configured
+  - the browser
+
+So what tools do we have at our disposal as Rails developers?
+
+```ruby
+config.time_zone = 'Eastern Time (US & Canada)' # rails application timezone
+config.active_record.default_timezone = :local  # controls the timezone datetime is converted to prior saving to DB
+```
+The most important one is the `config.time_zone` configuration in your __*config/application.rb*__ file. This sets the application timezone. ActiveRecord will help you convert from and to UTC and the timezone of your choice.
+
+Another one is `config.active_record.default_timezone` configuration in your __*config/application.rb*__ file, This controls in which timezone DB entries are treated at. It can be either UTC or Local. And UTC is default. If you set it to local, ActiveRecord will help you convert from and to system local and the timezone of your choice when setting and retrieving from the database.
```

Comment:
"Another one is **the** ..."

Reply:
"This controls which timezone DB entries are stored in."

Reply:
"It can be either `:utc` or `:local`, and UTC is the default."

Reply:
"to and from" sounds better than "from and to".

Reply:
> If you set it to local, ActiveRecord will help you convert from and to system local and the timezone of your choice when setting and retrieving from the database.

Technically this is true if you have this config set to UTC as well since AR will still ensure you get records back in the rails configured timezone (unless you make the query against the DB using something like `exec_query`). The only thing this changes is what timezone the records are in fact saved to the DB using.